### ISSUE TITLE
Design and implement a fancier heuristicblock data structure

### ISSUE THREAD

**Issue by pde** [EFForg/privacybadger#266]
Currently the httpRequestOriginFrequency data structure needs to store fragments of browsing history locally within the browser, in order to detect which 3rd party domains are tracking.  We should migrate to a fancier, higher-privacy data structure that prunes at the right time, uses HMAC fragments instead of origin names, and takes up less space.

**Comment by cooperq** [comment]
This was discussed in the bi-weekly developers meeting on 9/27. 

Threat model: A local attacker would be able to view a subset of cleared history (only visited domains, not urls or times).
This solution does not actually help against that since the attacker could still see if the users has  visited site n by taking H(n) and searching for it in privacy badger's database. 
what does this stop: an attacker who can view pb data but doesn't know how to hash strings.

other option: respect history clearing events, and remove domains from snitch map when they are removed from history. 
problem: this will break privacy badger if the user consistently clears their entire history.

**Comment by pde** [comment]
We definitely should do this.

**Comment by pde** [comment]
In particular, my instinct is that we should do the following two things:
- [ ] When domains get redlisted, remove the three first party entries that caused them to be redlisted.
- [ ] Maybe replace the first party domain names with a 1-2 hex digit HMAC slice. That's a simple special case of a bloom filter; it will make the algorithm a bit slower to block some trackers, but not terribly so.

**Comment by pjlsergeant** [comment]
>  respect history clearing events, and remove domains from snitch map when they are removed from history

I don't think this part is optional; I think this is a hard requirement

**Comment by pjlsergeant** [comment]
> Maybe replace the first party domain names with a 1-2 hex digit HMAC slice. That's a simple special case of a bloom filter; it will make the algorithm a bit slower to block some trackers, but not terribly so.

This is a great idea; it gives deniability, you can use a very very fast HMAC digest algorithm for it, and the only draw back I see is that once in x times you'll undercount unique domains.

**Comment by pde** [comment]
> I don't think this part is optional; I think this is a hard requirement if you are going to maintain the full domain name.

I agree that leaving cleartext history samples around is unacceptable if the user clears their history. But I've been wondering if 1-hex-digit records are acceptable in that case.

There are some corner cases where a small amount of information about browsing history may be genuinely inferrable from those records. For instance, if a third party domain is on a grand total of 5 first party domains across the Web, then `thirdparty-exampe.com : ["5", "A", "C"]` probably allows an attacker to infer three visited first parties from the five possibilities.  But such cases would be quite rare, and it isn't clear that these records are much worse than having `thirdparty-example.com` in our blocklist, which is more profoundly unavoidable.

**Comment by pjlsergeant** [comment]
The corner cases still leave you with a deniability factor that the plaintext domains don't. I think the single(or double)-character hash here is a clear winner.

**Comment by ghostwords** [comment]
I don't see a way to be notified of history-clearing events at this time without asking for the "history" permission, which will trigger a new permission warning, which seems unacceptable for Privacy Badger at this point (will lose a significant percentage of users in Chrome, and have a similar percentage stop upgrading in Firefox).
